zip2john: Refactor and support reading central directory

srett opened the pull request:
By default zip2john now reads the central directory and
processes files from there. This makes handling streaming
archives (where the local header is missing length information)
more robust. The old behavior of scanning the file from the
beginning is still available by passing the -s command line
switch.

Commit:
zip2john: Refactor and support reading central directory

By default zip2john now reads the central directory and
processes files from there. This makes handling streaming
archives (where the local header is missing length information)
more robust. The old behavior of scanning the file from the
beginning is still available by passing the -s command line
switch.

srett commented:
As discussed in #4054 

Apart from the central directory stuff I tried to make the naming of things a little more consistent and reduce the amount of duplicated code. ZIP64 support was also improved and tested.

I was not able to test $zip3$ support ("strong encryption" only used by pkzip IIRC), but there's no particular reason to assume it's broken (if it wasn't already broken before).

ghost commented:
The PR is **NOT** working with (my) AES files. 

Comparing the new hashes, the new local based hashes, and the old hashes (current implementation):
```
-rw-r--r--  1 root    root        2102 fev  8 12:45 local.txt
-rw-r--r--  1 root    root        2102 fev  8 12:37 new.txt
-rw-r--r--  1 claudio claudio  3133760 fev  8 12:42 old.txt
```
**[edited]**

As an example, new hash:
`2007.zip/1.JPG:$zip2$*0*3*0*1400000000000000d20d5064d77f0000*1e0f*0**e3304e71e82993282dbc*$/zip2$:1.JPG:2007.zip:/host/2007.zip
`

The old implementation
`2007.zip/1.JPG:$zip2$*0*3*0*1400000000000000d20d5064d77f0000*1e0f*81fbe*e3304e71e82993282dbc6a354501dd3dc218d4cd3b0572974fbd2b9c99204adaba1df ...[MUCH MORE]`

ghost commented:
One AES file to test: 
[2019-zipTest.zip](https://github.com/magnumripper/JohnTheRipper/files/4174639/2019-zipTest.zip)

Closes #4171 (the segfault itself is fixed)

srett commented:
Thanks for finding that one! It's kinda funny; the archive just has a 32bit central directory, but then the local file header contains a zip64 extended field saying the compressed and uncompressed length are both 0, which overwrites the proper 32bit values. I'll try to figure out what the right thing to to is here, otherwise I'll cobble together a simple workaround that seems to make sense. :-)

Commit:
zip2john: Workaround for archives with bogus ZIP64 data

srett commented:
I couldn't make too much sense from the APPNOTE.TXT on issues like this. I decided to ignore ZIP64 extended fields if they are zero and we already got something from the normal 32bit fields.

On another note, after a month of not looking at the code I forgot that I actually added "ZFILE" support. The big comment at the top of zip2john.c mentions it, john supports it, but zip2john never made use of that feature, so I added that too: Now, if the compressed data's size exceeds 100kb (arbitrarily chosen by me), it won't be inlined into the hash anymore, but reference the actual .zip file. So that would apply to your test file too. It means you need the original .zip file in addition to the hash to crack the password, so it might make sense to make this a command line switch... Since the inlined data is hex encoded, it's twice as large as the data in the .zip file. If you want to crack the password of a multi-gb file, it might actually make sense to use this.

srett commented:
I just noticed I somehow killed a return statement after an "if" without braces some time during refactoring. It's just affecting a debug print, but still... Should I just keep committing to this branch or create a new and clean squashed PR?

I'm applying the recommended astyle settings while I'm at it. Wasn't sure about it since a lot of the existing code didn't adhere to the style guide.

ghost commented:
We can squash while merging. But, the style stuff should be in its own commit (to make review easier). 

In fact, probably, you should forget the astyle stuff for now (it is easier for magnum to run it himself later).

Commit:
zip2john: exit() -> error()

Commit:
zip2john: Add missing return statement

ghost commented:
AES is cracking the proper password now. Thanks.

I see a side effect while testing the file attached to #4171. But we can't test it (we don't know the password).

ghost commented:
Everything seems ok to me.

I will wait for comments. I'll merge later today (to allow comprehensive testing).

magnumripper commented:
Awesome! I'll have a look-see very soon unless Claudio feels confident enough to merge it before that. Thanks!

srett commented:
Glad I could contribute something, john has been useful in the past. :)

A couple notes, remarks or ideas for the future:

- The zip format has a lot of redundancy and extensions (zip64) that allow for creating contradicting information (e.g. central vs. local data). zip2john isn't designed to verify that a zip file is malicious or suspicious, but it doesn't try too hard to even inform the user that something might be off. I added simple checks for length information mismatch, but that's about it. Well, just in this PR claudio supplied another variant of this happening.. :) But if there's ever going to be a rewrite of this tool I suggest to track file meta data from the central directory, local header and zip64 extended fields separately and at least output some warning.
- It might make sense to have the debug output not hidden behind a compile time flag but command line argument, for easier bug reporting by users, but this might be global john policy. Also the code for legacy pkzip encryption is rather verbose by default, for AES encryption nearly everything is hidden in ifdefs.
- As mentioned above, the threshold for not inlining the compressed data is fixed at 100kb. Making this a command line switch might come in handy, as you don't need the original zip file for cracking then, trading off size for convenience.
- The old version would either process an archive as AES or legacy. The new version handles this on a per-file basis. In theory, if you had an archive that contained both types of encrypted files, it would output both types of hashes, but I doubt this thing exists until deliberately created this way.
- Error checking/handling of file reads is inconsistent. The ```fgetXXLE()``` helpers pretty much prevent checking for errors, in some parts ```fread()```'s return code isn't checked, in some it is. Again, the question is how rigorous such a tool needs to check against these kind of failures, but if it doesn't even bail out on errors, worst case it might output bogus hashes that are uncrackable. I introduced ```xfread()``` which exits on error, but a bit later I realized that this might not be the best behavior if you pass the tool multiple .zip files at once, since it would not even try to read the remaining ones. I'll leave it up to someone else to decide here.
- It'd be great to have more test files, created by different tools and tool versions.
- There's central directory encryption and local file header masking. Didn't really check what they are, so not supported at all.

Attached is a file generated by zip4j, which contains a file >4GB and thus requires valid zip64 extended fields.
[aes-64bit.zip](https://github.com/magnumripper/JohnTheRipper/files/4178862/aes-64bit.zip)

magnumripper commented:
> the threshold for not inlining the compressed data is fixed at 100kb

Current implementations of all `*2john` tools should *always* inline, never point to a separate file. We started doing this after implementing support for arbitrary large input files a couple years back. A problem/drawback is that our input files are often twice as big as the actual input (eg. 4 GB hex data for a 2 GB archive file) but the pros win over the cons. Ideally we should always use Base64 instead of hex but for hysterical reasons we normally don't.

srett commented:
I see. I did this when working on ZIP64 since I ran out of memory with my test file (laptop has only 4GB, output hash was 8GB+). Since the comment mentioned it I assumed people just didn't get around to implementing it. So should it just be removed entirely again?

ghost commented:
>>  all *2john tools should always inline, never point to a separate file

> So should it just be removed entirely again?

It allows you to crack a password in the cloud, for example, without leaking or sharing the zip file itself. So, IMO, please do.

magnumripper commented:
> > > all *2john tools should always inline, never point to a separate file
> 
> > So should it just be removed entirely again?
> 
> It allows you to crack a password in the cloud, for example, without leaking or sharing the zip file itself. So, IMO, please do.

Not necessarily without leaking: The smallest file will be picked for inlining and if that file is the only file, we're basically leaking the whole archive.

We could keep a threshold for sure, but it should be much larger than 100 KB - perhaps 4 GB or something like that. Also, it should be a run-time option as opposed to compile-time. I'll look into it myself, @srett you don't need to change anything right now.

srett commented:
> Not necessarily without leaking: The smallest file will be picked for inlining and if that file is the only file, we're basically leaking the whole archive.

That logic only applies to the legacy encryption, where it tries to find the three smallest files. The AES code simply dumps hashes for every file found in the archive, so if you have a 4GB and a 2kb file, it will still dump everything to stdout, unless you use the ```-o``` option or grep the output. Assuming every file is encrypted with the same password, only outputting the smallest file might be another idea for the future here to make this even more of a corner case.  
But I guess for now, moving the threshold into the GB range and making it a cmdline switch should be sufficient.

magnumripper commented:
> > Not necessarily without leaking: The smallest file will be picked for inlining and if that file is the only file, we're basically leaking the whole archive.
> 
> That logic only applies to the legacy encryption, where it tries to find the three smallest files. The AES code simply dumps hashes for every file found in the archive, so if you have a 4GB and a 2kb file, it will still dump everything to stdout, unless you use the `-o` option or grep the output. Assuming every file is encrypted with the same password, only outputting the smallest file might be another idea for the future here to make this even more of a corner case.

Oh, right. We should fix this. Virtually all encrypted archives use the same password for every file. We should default to pick the smallest [usable] file and perhaps add some option to output a hash for every file (or simply have that situation handled with existing `-o` and multiple runs).

We might have the same "problem" with rar5tojohn - IIRC it outputs all files in the archive. But in that case size doesn't matter and (I think) we normally have a re-used salt, so it doesn't harm a lot.

magnumripper commented:
Squashed & merged manually. Many thanks!
